File: packages/agent-runtime/py/tests/test_write_lease.py
"""Process write lease: one writer per sqlite path; death releases; no TTL steal."""

from __future__ import annotations

import subprocess
import sys
import time
from pathlib import Path

import pytest
from steerable_agent_runtime.errors import StoreAlreadyOwnedError
from steerable_agent_runtime.storage import SqliteStorage
from steerable_agent_runtime.storage.write_lease import (
    acquire_write_lease,
    lock_path_for_db,
)

_HOLDER = r"""
import sys, time
from pathlib import Path
from steerable_agent_runtime.storage.write_lease import acquire_write_lease
lease = acquire_write_lease(sys.argv[1])
Path(sys.argv[2]).write_text("held", encoding="utf-8")
time.sleep(60)
"""


def _spawn_holder(db_path: Path, flag: Path) -> subprocess.Popen[str]:
    return subprocess.Popen(
        [sys.executable, "-c", _HOLDER, str(db_path), str(flag)],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )


def _wait_held(proc: subprocess.Popen[str], flag: Path, timeout: float = 5.0) -> None:
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if proc.poll() is not None:
            err = proc.stderr.read() if proc.stderr else ""
            raise AssertionError(
                f"holder exited {proc.returncode} before taking the lease: {err}"
            )
        if flag.exists() and flag.read_text(encoding="utf-8").strip() == "held":
            return
        time.sleep(0.05)
    err = proc.stderr.read() if proc.stderr else ""
    raise AssertionError(f"holder did not take the lease: {err}")


def test_lock_path_is_sibling_lock_file(tmp_path: Path) -> None:
    db = tmp_path / "sessions.db"
    assert lock_path_for_db(db) == tmp_path.resolve() / "sessions.lock"


def test_second_process_fails_loud(tmp_path: Path) -> None:
    db = tmp_path / "sessions.db"
    flag = tmp_path / "held"
    holder = _spawn_holder(db, flag)
    try:
        _wait_held(holder, flag)
        with pytest.raises(StoreAlreadyOwnedError, match="already owned"):
            acquire_write_lease(db)
        with pytest.raises(StoreAlreadyOwnedError):
            SqliteStorage(str(db))
        assert lock_path_for_db(db).is_file()
    finally:
        holder.kill()
        holder.wait(timeout=5)


def test_successor_opens_after_holder_is_killed(tmp_path: Path) -> None:
    db = tmp_path / "sessions.db"
    flag = tmp_path / "held"
    holder = _spawn_holder(db, flag)
    try:
        _wait_held(holder, flag)
        holder.kill()
        holder.wait(timeout=5)
    except BaseException:
        holder.kill()
        holder.wait(timeout=5)
        raise
    store = SqliteStorage(str(db))
    store.close()
    assert lock_path_for_db(db).is_file()


def test_release_does_not_delete_lock_file(tmp_path: Path) -> None:
    db = tmp_path / "sessions.db"
    lease = acquire_write_lease(db)
    lock = lock_path_for_db(db)
    assert lock.is_file()
    lease.release()
    assert lock.is_file()
    again = acquire_write_lease(db)
    again.release()
    assert lock.is_file()

File: packages/agent-runtime/rs/tests/test_write_lease.rs
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

use steerable_agent_runtime::{acquire_write_lease, lock_path_for_db, LeaseError};

struct TmpDir {
    path: PathBuf,
}

impl TmpDir {
    fn new(label: &str) -> Self {
        let nanos = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .expect("clock")
            .as_nanos();
        let path =
            std::env::temp_dir().join(format!("steerable-{label}-{}-{nanos}", std::process::id()));
        std::fs::create_dir_all(&path).expect("tmp dir");
        Self { path }
    }
}

impl Drop for TmpDir {
    fn drop(&mut self) {
        let _ = std::fs::remove_dir_all(&self.path);
    }
}

#[test]
fn holder_process_entry() {
    let Ok(db) = std::env::var("STEERABLE_WRITE_LEASE_DB") else {
        return;
    };
    let flag = std::env::var("STEERABLE_WRITE_LEASE_FLAG").expect("flag");
    let _lease = acquire_write_lease(&db).expect("holder lease");
    std::fs::write(&flag, "held").expect("flag");
    std::thread::sleep(Duration::from_secs(60));
}

fn spawn_holder(db: &Path, flag: &Path) -> std::process::Child {
    Command::new(std::env::current_exe().expect("test bin"))
        .arg("holder_process_entry")
        .arg("--exact")
        .env("STEERABLE_WRITE_LEASE_DB", db)
        .env("STEERABLE_WRITE_LEASE_FLAG", flag)
        .stdout(Stdio::null())
        .stderr(Stdio::piped())
        .spawn()
        .expect("spawn holder")
}

fn wait_held(child: &mut std::process::Child, flag: &Path) {
    let deadline = Instant::now() + Duration::from_secs(5);
    while Instant::now() < deadline {
        if let Some(code) = child.try_wait().unwrap() {
            let mut err = String::new();
            if let Some(mut stderr) = child.stderr.take() {
                use std::io::Read;
                let _ = stderr.read_to_string(&mut err);
            }
            panic!("holder exited {code} before taking the lease: {err}");
        }
        if flag.is_file() && std::fs::read_to_string(flag).unwrap_or_default().trim() == "held" {
            return;
        }
        std::thread::sleep(Duration::from_millis(50));
    }
    panic!("holder did not take the lease");
}

#[test]
fn lock_path_is_sibling_lock_file() {
    let dir = TmpDir::new("lock-path");
    let db = dir.path.join("sessions.db");
    assert_eq!(lock_path_for_db(&db), dir.path.join("sessions.lock"));
}

#[test]
fn second_process_fails_loud() {
    let dir = TmpDir::new("second-writer");
    let db = dir.path.join("sessions.db");
    let flag = dir.path.join("held");
    let mut holder = spawn_holder(&db, &flag);
    wait_held(&mut holder, &flag);
    let err = acquire_write_lease(&db).expect_err("second writer");
    match err {
        LeaseError::AlreadyOwned(owned) => {
            assert!(owned.to_string().contains("already owned"), "{owned}");
        }
        other => panic!("expected already owned, got {other}"),
    }
    assert!(lock_path_for_db(&db).is_file());
    holder.kill().ok();
    let _ = holder.wait();
}

#[test]
fn successor_opens_after_holder_is_killed() {
    let dir = TmpDir::new("successor");
    let db = dir.path.join("sessions.db");
    let flag = dir.path.join("held");
    let mut holder = spawn_holder(&db, &flag);
    wait_held(&mut holder, &flag);
    holder.kill().ok();
    holder.wait().unwrap();
    let mut lease = acquire_write_lease(&db).expect("successor");
    lease.release();
    assert!(lock_path_for_db(&db).is_file());
}

#[test]
fn release_does_not_delete_lock_file() {
    let dir = TmpDir::new("keep-lock");
    let db = dir.path.join("sessions.db");
    let mut lease = acquire_write_lease(&db).unwrap();
    let lock = lock_path_for_db(&db);
    assert!(lock.is_file());
    lease.release();
    assert!(lock.is_file());
    let mut again = acquire_write_lease(&db).unwrap();
    again.release();
    assert!(lock.is_file());
}
